lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,234 @@
+    """
+    A django model tracking persistent grades at the subsection level.
+
+    TODO: here are the other query patterns listed in the ticket that are not currently used. Should any of these
+    indices be built?
+        user_id, course_id, content_type, is_valid
+        course_id, edit-timestamp
+        edit-timestamp
+    """
+
+    class Meta(object):
+        index_together = [
+            ('user_id', 'usage_key')
+        ]
+
+        unique_together = (('user_id', 'usage_key'))
+
+    id = UnsignedBigIntAutoField(primary_key=True)  # pylint: disable=invalid-name
+    subtree_edited_date = models.DateTimeField('last content edit timestamp')
+    user_id = models.CharField(max_length=255)
+    earned_all = models.IntegerField()
+    possible_all = models.IntegerField()
+    earned_graded = models.IntegerField()
+    possible_graded = models.IntegerField()
+
+    course_id = CourseKeyField(max_length=255)
+    usage_key = UsageKeyField(max_length=255)
+    course_version = models.CharField('guid of latest course version', max_length=255)
+
+    #is_valid = models.BinaryField()  # Might be needed if doing async updates
+
+    visible_blocks = models.ForeignKey(VisibleBlocks)
+
+    def __unicode__(self):
+        """
+        Returns a string representation of this model.
+        """
+        return "PersistentSubsectionGrade for user {} in subsection {}".format(self.user_id, self.usage_key)
+
+    @classmethod
+    def save_grade(cls, **kwargs):
+        """
+        Wrapper for create_grade or update_grade, depending on which applies.
+        Takes the same arguments as both of thsoe methods.
+        """
+        try:
+            cls.update(**kwargs)
+        except cls.DoesNotExist:
+            cls.create(**kwargs)
+
+    @classmethod
+    def create(cls, **kwargs):
+        """
+        Instantiates a new model instance using the provided kwargs, formatted as follows:
+            user_id: "student12345"
+            usage_key: "block-v1:edX+BeAwesomeX+2016+type@subsection+block@f007ba11"
+            course_version: "deadbeef"
+            subtree_edited_date: "2016-08-01 18:53:24.354741"
+            earned_all: 6
+            possible_all: 12
+            earned_graded: 6
+            possible_graded: 8
+            visible_blocks: [<list of BlockRecord objects>]
+        """
+        visible_blocks_model = VisibleBlocks.create(blocks=kwargs['visible_blocks'])
+
+        model = cls.objects.create(
+            user_id=kwargs['user_id'],
+            course_id=kwargs['usage_key'].course_key,
+            usage_key=kwargs['usage_key'],
+            course_version=kwargs['course_version'],
+            subtree_edited_date=kwargs['subtree_edited_date'],
+            earned_all=kwargs['earned_all'],
+            possible_all=kwargs['possible_all'],
+            earned_graded=kwargs['earned_graded'],
+            possible_graded=kwargs['possible_graded'],
+            visible_blocks=visible_blocks_model,
+        )
```

Comment:
It will be clearer where extra logic is actually being applied if you replace all the `x=kwargs['x']` lines with a single `**kwargs` at the end.  That will also validate that extra arguments aren't getting dropped accidentally.